web/packages/teleterm/src/mainProcess/shell.ts

```diff
@@ -0,0 +1,132 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import fs from 'node:fs';
+import os from 'node:os';
+
+import which from 'which';
+
+import Logger from 'teleterm/logger';
+
+export interface Shell {
+  /**
+   * Shell identifier, for example, pwsh.exe or zsh
+   * (it doesn't have to be the same as the binary name).
+   * Used as an identifier in the app config or in a document.
+   * Must be unique.
+   * */
+  id: string;
+  /** Friendly name, for example, Windows PowerShell, zsh. */
+  friendlyName: string;
+  /** Shell executable, for example, C:\\Windows\system32\pwsh.exe, /bin/zsh. */
+  binPath: string;
+}
+
+export async function getAvailableShells(): Promise<Shell[]> {
+  switch (process.platform) {
+    case 'linux':
+    case 'darwin':
+      return getUnixShells();
+    case 'win32': {
+      return getWindowsShells();
+    }
+  }
+}
+
+export function getDefaultShell(availableShells: Shell[]): string {
+  const logger = new Logger();
+  switch (process.platform) {
+    case 'linux':
+    case 'darwin': {
+      const fallbackShell = 'bash';
+      const { shell } = os.userInfo();
+      const shellId = availableShells.find(
+        availableShell => availableShell.binPath === shell
+      )?.id;
+
+      if (!shellId) {
+        logger.error(
+          `Failed to read ${process.platform} platform default shell, using fallback: ${fallbackShell}.\n`
+        );
+
+        return fallbackShell;
+      }
+
+      return shellId;
+    }
+    case 'win32':
+      if (availableShells.find(shell => shell.id === 'pwsh.exe')) {
+        return 'pwsh.exe';
+      }
+      return 'powershell.exe';
+  }
+}
+
+async function getUnixShells(): Promise<Shell[]> {
+  const shells = await fs.promises.readFile('/etc/shells', {
+    encoding: 'utf-8',
+  });
+  return shells
+    .split(os.EOL)
+    .map(line => line.trim())
+    .filter(line => line && !line.startsWith('#'))
+    .map(binPath => {
+      const name = binPath.split('/').at(-1);
+      return {
+        id: name,
+        friendlyName: name,
+        binPath: binPath,
+      };
+    });
+}
+
+async function getWindowsShells(): Promise<Shell[]> {
+  const shells = await Promise.all(
+    [
+      {
+        id: 'powershell.exe',
+        friendlyName: 'Windows PowerShell (powershell.exe)',
+      },
+      {
+        id: 'pwsh.exe',
+        friendlyName: 'PowerShell (pwsh.exe)',
+      },
+      {
+        id: 'cmd.exe',
+        friendlyName: 'Command Prompt (cmd.exe)',
+      },
+      {
+        id: 'wsl.exe',
```

Comment:
With `wsl.exe` as the default shell, env vars are not set properly. Both the `TELEPORT` env vars as well as including tsh in Path. Connecting to a new kube cluster also doesn't properly show the welcome message and `KUBECONFIG` isn't set.

Reply:
>With wsl.exe as the default shell, env vars are not set properly. 

Ah that's right, shame on me for not checking that. Fortunately, the fix seems to be easy, we need to set [a special `WSLENV` variable](https://devblogs.microsoft.com/commandline/share-environment-vars-between-wsl-and-windows/#wsl-and-environment-variables-after-insider-17063). I only need to put all variables we want to set in it.

>Connecting to a new kube cluster also doesn't properly show the welcome message 

This is caused by ConPTY, I have a fix [here](https://github.com/gravitational/teleport/pull/45318).

Reply:
I fixed the problem with passing env vars to WSL, the only thing that doesn't work is using tsh from Path.
However, it seems to me that it doesn't work in regular PowerShell too, I will look into that tomorrow.

Reply:
This issue is a little weird and I'm not sure where is the problem exactly.
We prepend `tsh.exe` to the Path in https://github.com/gravitational/teleport/blob/ffaf66870d8d8fa6dc033325bc63714c06996af6/web/packages/teleterm/src/services/pty/ptyHost/buildPtyOptions.ts#L246
I noticed that if I change `Path` to `path` for win32, everything works (`tsh.exe` is available in every shell). But why? On master we use `Path` and tsh is prepended correctly.

Btw, there was already confusion in regard to Path on Windows in node.js https://github.com/nodejs/node/issues/20605.

Perhaps we should just switch to `path` on Windows?  When I run `node` in a Windows terminal and type `process.env`, there's a `path` property in the returned object.

Reply:
Did you check it on a packaged version of the app? Because I definitely did it only in the dev version which does not prepend anything to path.

Reply:
It's a little bit hard to test, because the installer also adds the folder with tsh to path. But you can package Connect and run it from win-unpacked.

Both PowerShells, cmd.exe and, surprisingly, custom bash from Git for Windows work fine.

```
> get-command tsh

CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Application     tsh.exe                                            0.0.0.0    C:\teleport\web\packages\teleterm\build\release\win-unp…
```

WSL does not work. The path with tsh is not appended to `$PATH`.

Reply:
>Did you check it on a packaged version of the app? Because I definitely did it only in the dev version which does not prepend anything to path.

Yes.

>But you can package Connect and run it from win-unpacked.

TIL about running the app from win-unpacked, thanks!

Ok, so after uninstalling the app, I ran the unpacked one, tsh wasn't available anywhere.
Then I changed `Path` to `path` and it started worked in all shells (WSL included, `which tsh.exe` returned a path). Still, I don't know why `Path` works on master :(

Reply:
After uninstalling Connect and running the app from win-unpacked, tsh works fine for me everywhere except in WSL and I did not have to change `Path` to `path` (I assume you mean `pathName` in `prependBinDirToPath`).

But then I looked at `$PATH` in WSL and I see that `/mnt/c/teleport/web/packages/teleterm/build/release/win-unpacked/resources/bin` is included there. Perhaps it's wrong for us to expect that `tsh` is going to just work in WSL? Shouldn't we need to execute `tsh.exe` in WSL instead? I don't remember what the conventions are.

Reply:
Yeah, under WSL we need to execute `tsh.exe` https://stackoverflow.com/a/47663269.

But now I'm really confused, why doesn't it work for me when `pathName` is `path`? I see that I don't have `/mnt/c/teleport/web/packages/teleterm/build/release/win-unpacked/resources/bin` in the `$PATH` (same for any other shell).
Maybe something is broken on my machine? I will try on a different one.

Reply:
I just remembered that someone posted their workaround for using `tsh` in WSL: https://github.com/gravitational/teleport/issues/9849#issuecomment-1472203018

As for the Path issues, nothing comes to my mind other than adding console.log and checking what's up. I suspect it's on our side, not necessarily the system, unless they changed how env vars work as a part of regular system update, which I guess is not that likely.

Reply:
Okay, I just tested this on a clean OS and it worked. I was able to use `tsh/tsh.exe` in powershell.exe. pwsh.exe, cmd.exe and `tsh.exe` in WSL. 
So it seems that something was wrong with my previous machine, but I have no idea what :( 
I think we can mark this issue as resolved.

Reply:
>I just remembered that someone posted their workaround for using tsh in WSL: https://github.com/gravitational/teleport/issues/9849#issuecomment-1472203018

I'm not sure if I see value in replacing `tsh.exe` with `tsh`. I think the official convention is to use `[app-name].exe` when running Win32 apps from WSL:
https://learn.microsoft.com/en-us/windows/wsl/filesystems#run-windows-tools-from-linux

Also, the users can install `tsh` in WSL, I think renaming `tsh.exe` to `tsh` would lead to confusion.

Reply:
>As for the Path issues, nothing comes to my mind other than adding console.log and checking what's up. I suspect it's on our side, not necessarily the system, unless they changed how env vars work as a part of regular system update, which I guess is not that likely.

OH, I'VE GOT IT!
The problem was that I have for some reason the "user" and "os" PATH env var called differently:
<img width="564" alt="image" src="https://github.com/user-attachments/assets/742d8574-fe77-4ef2-87c8-fc226bf86120">

This resulted in the PATH variable being called `path`, not `Path` in the `process.env` object. Node.js handles this by having its own setters and getters for PATH, so no matter what property you set, the single underlying value is updated.
Unfortunately, I missed that we merge many env vars by spreading the objects, so these setters/getters were lost.
As a result, in the  `env` object I had both `path` and `Path` (and the value of `Path` was ignored). 
To fix this problem, we only need to find a property with the correct name in the `env` object and update it.